invenio_rdm_records/services/services.py

```diff
@@ -131,3 +148,180 @@ def oai_result_item(self, identity, oai_record_source):
+    def create_guest_access_request_token(
+        self, identity, id_, data, expand=False, uow=None
+    ):
+        """Create an request token that can be used to create an access request."""
+        # TODO detect duplicates
+        # TODO proper permission checks
+        if authenticated_user in identity.provides:
+            raise PermissionDeniedError("request_guest_access")
+
+        record = self.record_cls.pid.resolve(id_)
+        if current_app.config.get("MAIL_SUPPRESS_SEND", False):
+            # TODO log warning
+            print("email sending has been disabled!")
+
+        access_token = AccessRequestToken.create(
+            email=data["email"],
+            full_name=data["full_name"],
+            message=data["message"],
+            record_pid=id_,
+            shelf_life=timedelta(hours=6),
+        )
+
+        verify_url = url_for(
+            "invenio_rdm_records_ext.verify_access_request_token",
+            **{"access-request-token": access_token.token},
+        )
+        uow.register(
+            EmailOp(
+                receiver=data["email"],
+                subject=_(
+                    "Access request for '%(record_title)s'",
+                    record_title=record.metadata["title"],
+                ),
+                html_body=_(
+                    (
+                        "Please verify your e-mail address via the following link "
+                        "in order to submit the access request: "
+                        '<a href="%(url)s">%(url)s</a>'
+                    ),
+                    url=verify_url,
+                ),
+                body=_(
+                    (
+                        "Please verify your e-mail address via the following link "
+                        "in order to submit the access request: %(url)s"
+                    ),
+                    url=verify_url,
+                ),
+            )
+        )
+
+        return {
+            "message": _("Verification link sent out, please check your e-mail inbox"),
+        }
+
+    @unit_of_work()
+    def create_guest_access_request(self, identity, token, expand=False, uow=None):
+        """Use the access request token to create an access request."""
+        # TODO proper permission checks
+        if authenticated_user in identity.provides:
+            raise PermissionDeniedError("request_guest_access")
+
+        access_token = AccessRequestToken.get_by_token(token)
+        if access_token is None:
+            return None
+
+        # TODO detect duplicates
+        access_token_data = access_token.to_dict()
+        record = self.record_cls.pid.resolve(access_token_data["record_pid"])
+        data = {
+            "payload": {
+                "permission": record.parent.access.settings.default_access_permission,
+                "email": access_token_data["email"],
+                "full_name": access_token_data["full_name"],
+                "token": access_token_data["token"],
```

Comment:
for good measure, we should probably create a new token here rather than re-using the existing one.